bobcat/bb-cumber/src/main/java/com/cognifide/qa/bb/cumber/Bobcumber.java

```diff
@@ -82,64 +81,33 @@ public Bobcumber(Class<?> clazz) throws InitializationError, IOException {
     super(clazz);
     storeFailedResults = clazz.isAnnotationPresent(StoreFailedResults.class);
     if (storeFailedResults) {
-      file = new File(clazz.getAnnotation(StoreFailedResults.class).value());
-      if (!file.exists()) {
-        PrintWriter writer = new PrintWriter(file, CharEncoding.UTF_8);
-        writer.close();
-      }
-      if (RERUN_FAILED_TESTS_CLASS_NAME.equals(clazz.getName()) && isFeatureFileEmpty()) {
-        noTestsToRerun = true;
-      }
+      featureFile = createFile(clazz.getAnnotation(StoreFailedResults.class).value());
+      String statisticsFilePath = properties.getProperty(ConfigKeys.BOBCAT_REPORT_STATISTICS_PATH);
+      statisticsFile = createFile(statisticsFilePath);
     }
+    isItFailedTestsRerun = clazz.isAnnotationPresent(FailedTestsRunner.class);
   }
 
   @Override
   public void run(RunNotifier notifier) {
-    if (noTestsToRerun) {
-      notifier.fireTestFinished(Description.EMPTY);
+    if (isItFailedTestsRerun && !canRerunFailedTests()) {
+      shutdownRerun(notifier);
       return;
```

Comment:
as we've discussed it is good to apply single return pattern so you could use else statement here with dedicated method managing rest of the logic